batchflow/models/torch/base.py

```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
 
-            # Transfer only the requested outputs to CPU
-            result = self.transfer_from_device(requested_outputs)
-            if single_output:
-                result = result[0]
+            chunked_outputs = []
+            for chunk_inputs, chunk_targets in zip(chunked_inputs, chunked_targets):
+                # Evaluate requested outputs
+                chunk_outputs = self._predict(inputs=chunk_inputs, targets=chunk_targets, outputs=outputs[:])
+                chunked_outputs.append(chunk_outputs)
+
+            # Aggregate the outputs from microbatches
+            result = self.aggregate_microbatches(outputs, chunked_outputs, single_output)
+
+            # Store info about current predict iteration
+            self.last_predict_info.update({
+                'amp': self.amp,
+                'batch_size': batch_size,
+                'microbatch_size': microbatch,
+                'steps': steps,
+                'outputs': outputs,
+            })
 
         finally:
             if lock:
                 self.model_lock.release()
         return result
 
+    def _predict(self, inputs, targets, outputs):
+        # Parse inputs
+        inputs = inputs[0] if len(inputs) == 1 else inputs
+        targets = targets[0] if len(targets) == 1 else targets
+
+        # Convert layer ids into LayerHooks
+        outputs = self.prepare_outputs(outputs)
+
+        output_container = {}
+        with torch.no_grad(), torch.cuda.amp.autocast(enabled=self.amp):
+            inputs = self.transfer_to_device(inputs)
+            predictions = self.model(inputs)
+
+            output_container['predictions'] = predictions
+
+            if targets:
+                targets = self.transfer_to_device(targets)
+                loss = self.loss(predictions, targets)
+                output_container['loss'] = loss
+
+        # Make all possible outputs
+        additional_outputs = self.compute_outputs(predictions=predictions)
+        output_container.update(additional_outputs)
+
+        # Log inner info
+        predictions_ = list(predictions) if isinstance(predictions, (tuple, list)) else [predictions]
+        self.last_train_info['predictions_shapes'] = [get_shape(item) for item in predictions_]
+        self.last_predict_info['available_outputs'] = list(output_container.keys())
+
+        # Retrieve requested outputs
+        requested_outputs = self.extract_outputs(outputs, output_container)
+
+        # Transfer only the requested outputs to CPU
+        return self.transfer_from_device(requested_outputs)
+
+
+    # Common utilities for train and predict
+    def split_into_microbatches(self, inputs, targets, microbatch, drop_last):
+        """ Split inputs and targets into microbatch-sized chunks. """
+        # Parse microbatch size
+        if microbatch:
+            if microbatch is True:
+                microbatch = self.microbatch
+            else:
+                microbatch = microbatch or self.microbatch
```

Comment:
Since `if microbatch` check above has passed, `or self.microbatch` will never be used

Reply:
Nice catch